File: python/433. Minimum Genetic Mutation.py
import collections

class Solution(object):
    def minMutation(self, start, end, bank):
        """
        :type start: str
        :type end: str
        :type bank: List[str]
        :rtype: int
        """
        bank = set(bank)
        if end not in bank: return -1
        start in bank and bank.remove(start)
        end in bank and bank.remove(end)
        que = []
        que.append(start)
        vis = collections.defaultdict(int)
        step = 0
        seq = 'ACGT'
        while que:
            sz = len(que)
            step += 1
            while sz:
                sz -= 1
                cur = que.pop(0)
                cur = list(cur)
                for i in range(8):
                    ch = cur[i]
                    for j in seq:
                        if j==ch:continue
                        cur[i] = j
                        str_cur = ''.join(cur)
                        if str_cur==end: return step
                        if str_cur not in bank: continue
                        bank.remove(str_cur)
                        que.append(str_cur)
                    cur[i] = ch

        return -1  

if __name__ == '__main__':
    start = "AACCGGTT"
    end = "AAACGGTA"
    bank = ["AACCGGTA","AACCGCTA","AAACGGTA"]
    su = Solution()
    print su.minMutation(start, end, bank)

File: go/src/leetcode/433/433. Minimum Genetic Mutation.go
/**
@author ZhengHao Lou
Date    2022/05/07
*/
package main

import "fmt"

/**
https://leetcode-cn.com/problems/minimum-genetic-mutation/
*/
func minMutation(start string, end string, bank []string) int {
	if start == end {
		return 0
	}
	has := make(map[string]bool)
	for _, b := range bank {
		has[b] = true
	}
	if !has[end] {
		return -1
	}
	
	que := []string{start}
	for step := 0; len(que) != 0; step++ {
		for i := len(que) - 1; i >= 0; i-- {
			s := que[0]
			que = que[1:]
			if s == end {
				return step
			}
			var keys []string
			for t := range has {
				var x int
				for i := range s {
					if s[i] != t[i] {
						x++
					}
				}
				if x == 1 {
					keys = append(keys, t)
					que = append(que, t)
				}
			}
			for _, t := range keys {
				delete(has, t)
			}
		}
	}
	
	return -1
}

func main() {
	fmt.Println(minMutation("AACCGGTT", "AACCGGTA", []string{"AACCGGTA"}))
	fmt.Println(minMutation("AACCGGTT", "AAACGGTA", []string{"AACCGGTA", "AACCGCTA", "AAACGGTA"}))
	fmt.Println(minMutation("AAAAACCC", "AACCCCCC", []string{"AAAACCCC", "AAACCCCC", "AACCCCCC"}))
	fmt.Println(minMutation("AACCGGTT", "AAACGGTA", []string{"AACCGATT", "AACCGATA", "AAACGATA", "AAACGGTA"}))
}
